library/proc_macro/src/bridge/arena.rs

```diff
@@ -60,48 +54,46 @@ impl Arena {
 
         let mut chunk = Box::new_uninit_slice(new_cap);
         let Range { start, end } = chunk.as_mut_ptr_range();
-        self.start.set(start);
-        self.end.set(end);
-        chunks.push(chunk);
+        self.start = start;
+        self.end = end;
+        self.chunks.push(chunk);
     }
 
     /// Allocates a byte slice with specified size from the current memory
     /// chunk. Returns `None` if there is no free space left to satisfy the
     /// request.
-    #[allow(clippy::mut_from_ref)]
-    fn alloc_raw_without_grow(&self, bytes: usize) -> Option<&mut [MaybeUninit<u8>]> {
-        let start = self.start.get().addr();
-        let old_end = self.end.get();
+    fn alloc_raw_without_grow(&mut self, bytes: usize) -> Option<&mut [MaybeUninit<u8>]> {
+        let start = self.start.addr();
+        let old_end = self.end;
         let end = old_end.addr();
 
         let new_end = end.checked_sub(bytes)?;
         if start <= new_end {
             let new_end = old_end.with_addr(new_end);
-            self.end.set(new_end);
+            self.end = new_end;
             // SAFETY: `bytes` bytes starting at `new_end` were just reserved.
             Some(unsafe { slice::from_raw_parts_mut(new_end, bytes) })
         } else {
             None
         }
     }
 
-    fn alloc_raw(&self, bytes: usize) -> &mut [MaybeUninit<u8>] {
+    fn alloc_raw(&mut self, bytes: usize) -> &mut [MaybeUninit<u8>] {
         if bytes == 0 {
             return &mut [];
         }
 
-        loop {
-            if let Some(a) = self.alloc_raw_without_grow(bytes) {
-                break a;
-            }
-            // No free space left. Allocate a new chunk to satisfy the request.
-            // On failure the grow will panic or abort.
-            self.grow(bytes);
+        if let Some(a) = self.alloc_raw_without_grow(bytes) {
+            // SAFETY: the lifetime is extended here, but then constrained again in `alloc_str`.
+            return unsafe { &mut *(a as *mut _) };
```

Comment:
Is it possible to remove the internal mutability without this change?
I'm not sure how it's related to the rest of the commit.